### ISSUE TITLE
Proposal: normalize the different syntaxes for pragmas

### ISSUE THREAD

**Issue by GULPF** [nim-lang/Nim#9543]
### Problem
There are many different ways to invoke pragmas, depending on the type of pragma. All these examples are valid Nim syntax and survives parsing:
```nim
 {.foo: "param".}       # 1
 {.foo "param".}        # 2
 {.foo bar: 1, baz: 2.} # 3
 {.foo(1, 2, 3).}       # 4
 {.foo[bar]: "param".}  # 5
```
Examples of usages:

1. Used by custom pragmas and many others, e.g the `deprecated` pragma: `{.deprecated: "reason".}`
2. Used by the `push` pragma, e.g `{.push deprecated.}`
3. Used by the `push` pragma, e.g `{.push deprecated: "reason".}`
4. Used by custom pragmas
5. Used by the `hint` pragma, e.g `{.hint[LineTooLong]: off.}`

We should strive to normalize the syntax for pragmas because the current situation is confusing and doesn't bring any benefit. When calling a routin in other contexts Nim allows syntax `1`, `2` and `4`.

### Proposal
- Special pragmas like `push`, `hint`, and `pragma`should be reworked to use syntax `1`, `2` and `4`. Their current syntax should be deprecated then removed.
- Syntax `1`, `2` and `4` should be supported for all types of pragmas. Currently pragmas on routines like `deprecated` must use `1`, but user-defined pragmas on types can use `1` or `4`.
- The AST for pragma calls should ideally match the AST for normal calls. This is already true for `4`, but not for `1` and `2`. This would be a breaking change however.

**Comment by cooldome** [comment]
I like the way of pragmas mimicing procs. There many ways to call a proc and there should be as many ways to call a pragma. Couple of pragmas can be reworked I agree, proposal needs a full list specified. Passing arbitrary expression to pragma is actually a good thing in my opinion

**Comment by GULPF** [comment]
> I like the way of pragmas mimicing procs. There many ways to call a proc and there should be as many ways to call a pragma. 

To clarify, I agree with this. It's not the case currently, for example  `{.foo, bar.}` and `{.foo bar.}` is equivalent.

> Passing arbitrary expression to pragma is actually a good thing in my opinion

My complaint wasn't about passing expressions to pragmas. Look at the example, the expression isn't being passed anywhere.

**Comment by Araq** [comment]
> The AST for pragma calls should ideally match the AST for normal calls. This is already true for 4, but not for 1 and 2. This would be a breaking change however.

Yeah but I want this too and it's worth it. I doubt much code out there is affected, many people don't know about these optional commas. Plus we can have a deprecation period of course.

**Comment by timotheecour** [comment]
@GULPF will that also fix https://github.com/nim-lang/Nim/issues/2331 ? 
seems to me `{.error:"FooBarFooBar FooBarFooBarFooBar "FooBarFooBar" FooBarFooBarFooBarFooBar123".}` should be a syntax error if we're gonna normalize the syntax for pragma

**Comment by GULPF** [comment]
@timotheecour Yes, it would be a syntax error with this proposal.
